trie/branches.py

```diff
@@ -0,0 +1,196 @@
+    elif nodetype == BRANCH_TYPE:
+        if not keypath:
+            raise InvalidKeyError("Key too short")
+        if keypath[:1] == BYTE_0:
+            yield node
+            yield from _get_branch(db, left_child, keypath[1:])
+        else:
+            yield node
+            yield from _get_branch(db, right_child, keypath[1:])
+    else:
+        raise Exception("Invariant: unreachable code path")
+
+
+def if_branch_valid(branch, root_hash, key, value):
+    # value being None means the key is not in the trie
+    if value is not None:
+        validate_is_bytes(key)
+    # branch must not be empty
+    assert branch
+    for node in branch:
+        validate_is_bin_node(node)
+
+    db = {keccak(node): node for node in branch}
+    assert BinaryTrie(db=db, root_hash=root_hash).get(key) == value
+    return True
+
+
+def get_trie_nodes(db, node_hash):
+    """
+    Get full trie of a given root node
+    """
+    return tuple(_get_trie_nodes(db, node_hash))
+
+
+def _get_trie_nodes(db, node_hash):
+    if node_hash in db:
+        node = db[node_hash]
+    else:
+        raise StopIteration
+    nodetype, left_child, right_child = parse_node(node)
+    if nodetype == KV_TYPE:
+        yield node
+        yield from get_trie_nodes(db, right_child)
+    elif nodetype == BRANCH_TYPE:
+        yield node
+        yield from get_trie_nodes(db, left_child)
+        yield from get_trie_nodes(db, right_child)
+    elif nodetype == LEAF_TYPE:
+        yield node
+    else:
+        raise Exception("Invariant: unreachable code path")
+
+
+def get_witness(db, node_hash, key):
+    """
+    Get all witness given a keypath prefix.
+    Include
+
+    1. witness along the keypath and
+    2. witness in the subtrie of the last node in keypath
+    """
+    validate_is_bytes(key)
+
+    return tuple(_get_witness(db, node_hash, encode_to_bin(key)))
+
+
+def _get_witness(db, node_hash, keypath):
+    if not keypath:
+        yield from get_trie_nodes(db, node_hash)
+    if node_hash in db:
+        node = db[node_hash]
+    else:
+        raise StopIteration
+    nodetype, left_child, right_child = parse_node(node)
+    if nodetype == LEAF_TYPE:
+        if keypath:
+            raise InvalidKeyError("Key too long")
+    elif nodetype == KV_TYPE:
+        if len(keypath) < len(left_child) and left_child[:len(keypath)] == keypath:
+            yield node
+            yield from get_trie_nodes(db, right_child)
+        elif keypath[:len(left_child)] == left_child:
+            yield node
+            yield from _get_witness(db, right_child, keypath[len(left_child):])
+        else:
+            yield node
+    elif nodetype == BRANCH_TYPE:
+        if keypath[:1] == BYTE_0:
+            yield node
+            yield from _get_witness(db, left_child, keypath[1:])
+        else:
+            yield node
+            yield from _get_witness(db, right_child, keypath[1:])
+    else:
+        raise Exception("Invariant: unreachable code path")
```

Comment:
`get_witness` was renamed from `get_prefix_witness`, correct? I actually like the older naming better, because it reflects the fact that this method is supposed to return a witness that covers all key/value pairs that have a given prefix.

Reply:
My feeling was that `get_prefix_witness` might make people think that it's the *prefix of witness*. So I take *prefix* out of the name and explain it in docstring:
```
Get all witness given a keypath prefix.  Include
1. witness along the keypath and
2. witness in the subtrie of the last node in keypath
```

what about `get_witness_by_key_prefix`?